### PR TITLE
[WIP][16.0][MIG] shopinvader_api_quotation

### PR DESCRIPTION
**Pull request by matthieusaison**
Migrate shopinvader_quotation to shopinvader_api_quotation with fastapi

### CODE REVIEW

File: shopinvader_api_quotation/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+{
+    "name": "Shopinvader Api Quotation",
+    "summary": "Shopinvader Quotation",
+    "version": "16.0.1.0.1",
+    "category": "e-commerce",
+    "website": "https://github.com/shopinvader/odoo-shopinvader",
+    "author": "Akretion",
+    "license": "AGPL-3",
+    "depends": ["crm", "fastapi", "shopinvader_schema_sale", "shopinvader_api_cart"],
```
Comment (sebastienbeau) [inline shopinvader_api_quotation/__manifest__.py]: this module should depend on shopinvader_api_security_sale and  shopinvader_schema_sale, sale_cart (crm, shopinvader_api_cart can be removed)

File: shopinvader_api_quotation/models/sale_order.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
+# Copyright 2017-2018 Akretion (http://www.akretion.com).
+# Copyright 2021 Camptocamp (http://www.camptocamp.com)
+# @author Benoît GUILLOT <benoit.guillot@akretion.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+
+from odoo import fields, models
+
+# from odoo.exceptions import UserError
```
Comment (sebastienbeau) [inline shopinvader_api_quotation/models/sale_order.py]: remove comment

File: shopinvader_api_quotation/routers/quotation.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    count, orders = env["shopinvader_api_quotation.quotations_router.helper"]._search(
+        paging, params
+    )
+    return PagedCollection[Sale](
+        count=count,
+        items=[Sale.from_sale_order(order) for order in orders],
+    )
+
+
+class ShopinvaderApiQuotationquotationsRouterHelper(models.AbstractModel):
```
Comment (sebastienbeau) [inline shopinvader_api_quotation/routers/quotation.py]: see : https://github.com/shopinvader/odoo-shopinvader/blob/16.0/shopinvader_api_sale/routers/sales.py

File: shopinvader_api_quotation/routers/quotation.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+from odoo.addons.base.models.res_partner import Partner as ResPartner
+from odoo.addons.extendable_fastapi.schemas import PagedCollection
+from odoo.addons.fastapi.dependencies import (
+    authenticated_partner,
+    authenticated_partner_env,
+    paging,
+)
+from odoo.addons.fastapi.schemas import Paging
+from odoo.addons.fastapi.utils import FilteredDomainAdapter
```
Comment (sebastienbeau) [inline shopinvader_api_quotation/routers/quotation.py]: from odoo.addons.shopinvader_filtered_model.utils import FilteredModelAdapter

File: shopinvader_api_quotation/routers/quotation.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+@quotation_router.get("/quotations/{quotation_id}/confirm", status_code=200)
+def confirm_quotation(
+    env: Annotated[Environment, Depends(authenticated_partner_env)],
+    partner: Annotated[ResPartner, Depends(authenticated_partner)],
+    quotation_id: int | None = None,
+) -> None:
+    order = env["shopinvader_api_quotation.quotations_router.helper"]._confirm(
+        quotation_id
+    )
+    # env["sale.order"]._confirm(order)
```
Comment (sebastienbeau) [inline shopinvader_api_quotation/routers/quotation.py]: Todo remove comment

File: shopinvader_api_quotation/routers/quotation.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+) -> PagedCollection[Sale]:
+    count, orders = env["shopinvader_api_quotation.quotations_router.helper"]._search(
+        paging, params
+    )
+    return PagedCollection[Sale](
+        count=count,
+        items=[Sale.from_sale_order(order) for order in orders],
+    )
+
+
```
Comment (sebastienbeau) [inline shopinvader_api_quotation/routers/quotation.py:68]: TODO add update

see: https://github.com/shopinvader/odoo-shopinvader/blob/b42cf29964111c8af6dad796d695342a5bfc62d3/shopinvader_api_cart/routers/cart.py#L57


The aim is to be able to update the client_order_ref

{"customer_ref": "PO_XXXX"}


spec:

```
@router.post("/quotations/{quotation_id}")
def update_quotation(
    data: QuotationUpdateInput,
    env: Annotated[api.Environment, Depends(authenticated_partner_env)],
    partner: Annotated["ResPartner", Depends(authenticated_partner)],
    quotation_id: int,
):

=> Sale.from_quotation(quotation)

```

File: shopinvader_api_quotation/routers/__init__.py [hunk 7]
```diff
@@ -0,0 +1 @@
+from . import quotation
```
Comment (sebastienbeau) [inline shopinvader_api_quotation/routers/__init__.py:1]: TODO add a cart router


Extend Sale pydantic model
```
{
    "available_for_quotation": True,
    "shop_only_quotation": True/ False, (if we have a product shop_only_quotation in a sale order)
}
```


exemple: https://github.com/shopinvader/odoo-shopinvader/blob/b42cf29964111c8af6dad796d695342a5bfc62d3/shopinvader_api_cart/schemas/sale.py#L8



```
@router_cart.post("{uuid}/request_quotation")
def request_quotation(
    uuid: uuid
    env
    partner
)

return Sale.from_sale_order(order)

```

Comment (matthieusaison) [reply shopinvader_api_quotation/routers/__init__.py:1]: How is compute available_for_quotation? 
is it always True?

File: shopinvader_api_quotation/models/product_template.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+# @author Benoît GUILLOT <benoit.guillot@akretion.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models
+
+
+class ProductTemplate(models.Model):
+    _inherit = "product.template"
+
+    shop_only_quotation = fields.Boolean(
```
Comment (sebastienbeau) [inline shopinvader_api_quotation/models/product_template.py]: On V16 the following code will not work anymore (due to some ORM cleanup)
We should replace the boolean on the product_template by a "selection field" with the following value

```
   ("all_variant", "All Variant"),
   ("manually_on_variant", "Manually on Variant"),
   ("never", "Never"),
```


Then the inverse method will do the following

all_variant => all variant to True
manually_on_variant => do nothing
never => all variant to False

The compute method should do the following
all variant to True => all_variant
mix => manually_on_variant
all variant to False => never

So doing a recompute will be stable and the test will be not brokend anymore

Comment (sebastienbeau) [reply shopinvader_api_quotation/models/product_template.py]: we should also rename the field to "shop_only_quotation_mode" so we do not have "name" conflict between product and template

**Comment by sebastienbeau** [comment]
After thinking twice, I think it will be great to create a base module "sale_quotation" will all the "odoo logic" (everything we have in the models + views and tests on product).

So shopinvader_quotation (use with restapi) and shopinvader_api_quotation can depend on it.

Also in a long term we can improve the flow of sale_quotation to add more feature so it can be totaly usable without any frontend implementation.

File: shopinvader_api_quotation/views/product_view.xml [hunk 9]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<odoo>
+
+    <record id="view_product_template_form" model="ir.ui.view">
+        <field name="model">product.template</field>
+        <field name="inherit_id" ref="shopinvader.view_product_template_form" />
```
Comment (sebastienbeau) [inline shopinvader_api_quotation/views/product_view.xml]: Todo move this field elsewhere

File: shopinvader_api_quotation/routers/cart.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+from odoo import api
+
+from odoo.addons.base.models.res_partner import Partner as ResPartner
+from odoo.addons.fastapi.dependencies import (
+    authenticated_partner,
+    authenticated_partner_env,
+)
+from odoo.addons.shopinvader_schema_sale.schemas.sale import Sale
+
+cart_router = APIRouter(tags=["carts"])
```
Comment (sebastienbeau) [inline shopinvader_api_quotation/routers/cart.py]: Do not redefine a new cart router
Use the existing one

```suggestion
from odoo.addons.shopinvader_api_cart.routers import cart_router
```
 from odoo.addons.shopinvader_api_cart.routers import cart_router

File: shopinvader_api_quotation/schemas/sale.py [hunk 11]
```diff
@@ -0,0 +1,25 @@
+from extendable_pydantic import StrictExtendableBaseModel
```
Comment (sebastienbeau) [inline shopinvader_api_quotation/schemas/sale.py:1]: The `__init__.py__` is missing in the directory schemas

File: shopinvader_api_quotation/routers/quotation.py [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+    _name = "shopinvader_api_quotation.quotations_router.helper"
+    _inherit = "shopinvader_api_sale.sales_router.helper"
+
+    def _get_domain_adapter(self):
+        return [
+            ("partner_id", "=", self.partner.id),
+            ("typology", "=", "quotation"),
+        ]
+
+    def _confirm(self, quotation):
```
Comment (sebastienbeau) [inline shopinvader_api_quotation/routers/quotation.py:94]:
```suggestion
    def _confirm(self, quotation_id):
```

File: shopinvader_api_quotation/routers/quotation.py [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+        order = self._get(quotation)
+        order.action_confirm()
+        return order
+
+    def _update(self, quotation, data):
```
Comment (sebastienbeau) [inline shopinvader_api_quotation/routers/quotation.py:99]:
```suggestion
    def _update(self, quotation_id, data):
```

**Review by sebastienbeau — CHANGES_REQUESTED** [review]

**Comment by matthieusaison** [comment]
I close this PR, and continue here  for a more readable work: https://github.com/shopinvader/odoo-shopinvader/pull/1471

**Pull request closed without merging**